polkadot/runtime/parachains/src/inclusion/mod.rs

```diff
@@ -495,49 +503,55 @@ impl<T: Config> Pallet<T> {
 	/// Returns a `Vec` of `CandidateHash`es and their respective `AvailabilityCore`s that became
 	/// available, and cores free.
 	pub(crate) fn update_pending_availability_and_get_freed_cores<F>(
-		expected_bits: usize,
 		validators: &[ValidatorId],
 		signed_bitfields: SignedAvailabilityBitfields,
 		core_lookup: F,
 	) -> Vec<(CoreIndex, CandidateHash)>
 	where
 		F: Fn(CoreIndex) -> Option<ParaId>,
 	{
-		let mut assigned_paras_record = (0..expected_bits)
-			.map(|bit_index| core_lookup(CoreIndex::from(bit_index as u32)))
-			.map(|opt_para_id| {
-				opt_para_id.map(|para_id| (para_id, PendingAvailability::<T>::get(&para_id)))
-			})
-			.collect::<Vec<_>>();
-
 		let now = <frame_system::Pallet<T>>::block_number();
+		let threshold = availability_threshold(validators.len());
+
+		let mut paras_made_available = BTreeSet::new();
+
 		for (checked_bitfield, validator_index) in
 			signed_bitfields.into_iter().map(|signed_bitfield| {
 				let validator_idx = signed_bitfield.validator_index();
 				let checked_bitfield = signed_bitfield.into_payload();
 				(checked_bitfield, validator_idx)
 			}) {
 			for (bit_idx, _) in checked_bitfield.0.iter().enumerate().filter(|(_, is_av)| **is_av) {
-				let pending_availability = if let Some((_, pending_availability)) =
-					assigned_paras_record[bit_idx].as_mut()
-				{
-					pending_availability
+				let core_index = CoreIndex(bit_idx as u32);
+				if let Some(para_id) = core_lookup(core_index) {
+					<PendingAvailability<T>>::mutate(&para_id, |candidates| {
+						if let Some(candidates) = candidates {
+							for (candidate_idx, candidate) in candidates.iter_mut().enumerate() {
+								if candidate.core == core_index {
+									// defensive check - this is constructed by loading the
+									// availability bitfield record, which is always `Some` if
+									// the core is occupied - that's why we're here.
+									if let Some(mut bit) = candidate
+										.availability_votes
+										.get_mut(validator_index.0 as usize)
+									{
+										*bit = true;
```

Comment:
Yes, I agree, this will be very slow. While your suggestion makes sense, I'd argue that it might not be wise to allocate such large quantity of memory here. Maybe it would be better to put the commitments in separate storage.

Reply:
But, `PendingAvailabilityCommitments` is already a separate storage item, I guess in-memory should work fine.

Reply:
> I would suggest rewriting this into a small in-memory procedure which just translates per validator bitfields into votes per core and use that end result to write once into PendingAvailability.

I think this is a good suggestion